Fakify leaf of FunctionalTensor

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107337
* __->__ #107062
* #107042

This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ydwu4 committed:
Fakify leaf of FunctionalTensor

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1512,6 +1513,17 @@ def update_dim2constraint(dim, constraint_range):
 
+def replace_leaf_with_fake_tensor_and_record(e, tx, ignore_subclass=False, *, source: Optional[Source], is_tensor: bool):
+    if torch._is_functional_tensor(e):
+        from torch._C._functorch import _wrap_functional_tensor, _unwrap_functional_tensor, current_level
+
+        reapply_views = torch._C._functionalization_reapply_views_tls()
+        unwrap_e = _unwrap_functional_tensor(e, reapply_views)
+        replaced_e = replace_leaf_with_fake_tensor_and_record(unwrap_e, tx, ignore_subclass, source=source, is_tensor=is_tensor)
+        assert torch._is_functional_tensor(replaced_e) and isinstance(replaced_e, FakeTensor)
+        return replaced_e
```
Uh I'm confused. I would expect that we do the following:
- FunctionalTensor(to_fake(unwrap(e))). I do not see where we are constructing a new FunctionalTensor.

ydwu4 replied:
I think it's because we didn't disable the functionalize key, so in wrap_to_fake_tensor_and_record, it eventually calls empty_strided gives us a FunctionalTensor. It further wraps the FunctionalTensor with FakeTensor.

Is there any difference between FakeTensor(FunctionalTensor) vs FunctionalTensor(FakeTensor)? I mean from dispatcher's point of view, they should be the same?

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"


This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_subclasses.py:
```diff
@@ -128,6 +128,26 @@ def f(x):
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(cnt.op_count, 1)
+
+        ff = torch.func.functionalize(f)
+        ff_out = ff(x)
```
Just to check: we only care about functorch.functionalize() here, and not the AOTAutograd variant of funcitonalization? Or do we need both to be fakified?

zou3519 replied:
Hmm, the answer is probably both, AOTAutograd might be able to see torch.cond if implementation of any operator directly uses it

ydwu4 replied:
I can add some tests for AOTAutograd version of functionalization

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -523,6 +530,24 @@ def __call__(
                 # instrumentation will see the meta conversions and the
                 # tests all break so we just exclude this.  In any case
                 # the to conversion isn't really right anyhow.
+
```
FYI @bdhirsh - we're doing something that morally should just be adding a tensor_unflatten / tensor_flatten to FunctionalTensorWrapper, but since those APIs don't exist yet we're just hardcoding this into the fakeification logic. Curious if you think this is the right place to toss it or if you were putting the tensor_unflatten/tensor_flatten handling somewhere else

bdhirsh replied:
Is the goal of this PR specifically to get `torch.compile` working on code that runs inside of a `functorch.functionalize()` call? Or is it to more generically support functionalization wherever its used in the stack (functionalization kernel for torch.cond, and functionalization that runs in AOTAutograd).

I'm asking because AOTAutograd is (hopefully) soon going to start using a python version of `FunctionalTensorWrapper` ([PR](https://github.com/pytorch/pytorch/pull/106405/files#diff-a578c8c54653bc4c7b0775ad273154ac7e9ece893f6bf2fda1874c088b16072aR1)). Although I haven't given it a flatten/unflatten API (mostly because in AOTAutograd, I don't actually **want** to flatten it).

I wonder if you think it'd be worth just handling this logic for the python `FunctionalTensor` and not bothering for the C++ variant. That'd probably require updating torch.cond's functionalization rule to use the python flavor. And.. I'm also not sure if we can easily migrate `functorch.functionalize`'s version to use the python flavor.

But either way, I think this probably a reasonable place to put this code.

zou3519 commented on test/dynamo/test_subclasses.py:
```diff
@@ -128,6 +128,26 @@ def f(x):
+        x = torch.randn([3, 4])
+        cnt = torch._dynamo.testing.CompileCounter()
+
+        @torch.compile(backend=cnt, fullgraph=True)
+        def f(x):
+            return torch.sin(x)
```
- Can we put some mutable operations into the graph and then do an expect test on the graph that sent to the backend? I expect we will see all the mutable operations in the graph
- Can we assert that the sample inputs passed to the backend are actually FunctionalTensorWrapper?
- Could we check that torch.compile(backend='eager', fullgraph=True) works? (this is the case that we care about for torch.cond)

ydwu4 replied:
Good idea!

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -523,6 +530,24 @@ def __call__(
+                if torch._is_functional_tensor(t) and t.device.type != "lazy":
+                    reapply_views = torch._C._functionalization_reapply_views_tls()
+                    unwrap_t = _unwrap_functional_tensor(t, reapply_views)
+                    st = peek_interpreter_stack()
```
could we assert that st is a FunctionalizeInterpreter?

zou3519 reviewed (commented):
Code looks good to me, I want some more tests

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"


This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_subclasses.py:
```diff
@@ -128,6 +143,74 @@ def f(x):
+
+        self.assertEqual(cnt.frame_count, 2)
+        self.assertEqual(cnt.op_count, 6)
+        self.assertEqual(len(backend.graphs), 2)
```
Uh, why are there two graphs? we even have fullgraph=True

zou3519 replied:
Oh I see we called the function again and the counter counts that

ydwu4 replied:
Yeah, the guard fails due to different input tensor type. So the user compiler is invoked again

zou3519 commented on test/dynamo/test_subclasses.py:
```diff
@@ -128,6 +143,74 @@ def f(x):
+        ff = torch.func.functionalize(f)
+        ff_out = ff(x_clone)
+
+        self.assertEqual(cnt.frame_count, 2)
+        self.assertEqual(cnt.op_count, 6)
+        self.assertEqual(len(backend.graphs), 2)
+        self.assertEqual(len(backend.example_inputs), 2)
+        actual = normalize_gm(backend.graphs[1].print_readable(print_output=False))
```
I'd recommend not checking cnt and backend again, because we already did it one time? Ditto below

zou3519 replied:
Gotcha, thanks for the explanation. I'm convinced, let's leave them

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -523,6 +531,45 @@ def __call__(
+                                dynamic_dims=dynamic_dims,
+                                constraint_dims=constraint_dims,
+                            )
+                        return _wrap_functional_tensor(fake_t, current_level())
```
That's a good point. I'd just recommend detecting it and erroring out.

@bdhirsh are the considerations the same for fakeifying wrapper Tensor subclasses? Some of them might be views of each other and fakeification should preserve the view relationships?

bdhirsh replied:
> are the considerations the same for fakeifying wrapper Tensor subclasses? Some of them might be views of each other and fakeification should preserve the view relationships?

If a wrapper subclass holds onto an inner tensor that is secretly an alias of another input to the graph, that isn't supported (so we'll probably need some tool to detect and error out when that happens).

If one of the inner tensors is a view (of some ambient tensor that isn't part of the compiled region), then... I think this should work, but it's worth adding tests for (i'll add this to my list of tests to add for the subclass work). We just need to make sure that AOTAutograd notices the input mutation at the autograd level, so when it re-applies the mutation, the autograd engine is aware of it.

ydwu4 replied:
Yeah, erroring out seems good to me.

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"


This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"


This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Fakify leaf of FunctionalTensor"


This PR allows dynamo to fakify FunctionalTensorWrapper by unwrapping, replacing and wrapping again for FunctionalTensorWrapper so that FunctionalTensorWrapper can be passed in as input for dynamo.optimize and we can support something like this
```python
ff = torch.func.functionalize(f)
torch.compile(ff)(x)
```

This PR didn't follow the \_\_tensor_flatten\_\_ and \_\_tensor_unflatten\_\_ protocol right now because we're not sure the plan of doing that for FunctionalTensorWrapper (it's implemented in C++). 

**Test Plan:**
Add a new test.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -m "This appears to have broken the test TestDTensorCompile.test_dtensor_fullgraph.  Probably a land race" -c landrace

ZainRizvi commented:
See logs here for more details:
https://hud.pytorch.org/pytorch/pytorch/commit/6cb0128c8a07d626ab84516df3c9727943469d49

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot revert -m "This appears to have broken the test TestDTensorCompile.test_dtensor_fullgraph."

ydwu4 commented:
@pytorchbot revert -m "This appears to have broken the test TestDTensorCompile.test_dtensor_fullgraph. Probably a land race" -c landrace
